### PR TITLE
Avoid repeated binary cache server operations.

### PR DESCRIPTION
**Pull request by BillyONeal**
We recently have run into problems in our normal CI infrastructure when we increased the size of the VMs where all VMs tried to fetch the same nonexistent entries over and over and caused failures. To alievate that problem, this changes binary caching to not issue server I/Os for status that vcpkg already knows.

To that end, this is a substantial redesign of the IBinaryProvider mechanism to give this global knowledge of which things have been restored from cache, such that MergedBinaryProvider is replaced with class BinaryCache, and the functions that were 'masaging' the interface to IBinaryProvider are put in that type. Moreover, NullBinaryProvder is an empty binary cache.

Incidential fixes:
* Different commands were inconsistent in how they configured the binary cache, in particular manifest mode and normal mode installs treated --download-only differently.
* The different IBinaryProviders that make a temp zip file all use the same zip file name.

### CODE REVIEW

File: include/vcpkg/binarycaching.h [hunk 1]
```diff
@@ -9,53 +9,131 @@
...
-        virtual RestoreResult try_restore(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) = 0;
-
-        /// Called upon a successful build of `action`
-        virtual void push_success(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) = 0;
-
-        /// <summary>Gives the BinaryProvider an opportunity to batch any downloading or server communication for
-        /// executing `plan`.</summary>
-        /// <remarks>Must only be called once for a given binary provider instance</remarks>
-        /// <param name="actions">InOut vector of actions to be prefetched</param>
+        /// Prerequisite: action has a package_abi()
```
Comment (BillyONeal) [inline include/vcpkg/binarycaching.h:79]: Note that class BinaryCache blocks non package_abi'd things

File: src/vcpkg/dependencies.cpp [hunk 2]
```diff
@@ -472,14 +472,14 @@ namespace vcpkg::Dependencies
...
+        return p && !p->package_abi.empty();
     }
     Optional<const std::string&> InstallPlanAction::package_abi() const
     {
-        if (!abi_info) return nullopt;
-        if (abi_info.get()->package_abi.empty()) return nullopt;
-        return abi_info.get()->package_abi;
+        const auto p = abi_info.get();
+        if (!p || p->package_abi.empty()) return nullopt;
+        return p->package_abi;
```
Comment (strega-nil-ms) [inline src/vcpkg/dependencies.cpp:482]: Is there a reason these changes are here? They're fine, but they feel unrelated.

File: src/vcpkg/binarycaching.cpp [hunk 3]
```diff
@@ -903,179 +906,490 @@ namespace
...
+
+    CacheStatus::CacheStatus() noexcept : m_status(CacheStatusState::unknown), m_known_unavailable_providers() { }
+
+    CacheStatus::CacheStatus(const CacheStatus& other) : m_status(other.m_status)
+    {
+        switch (other.m_status)
+        {
+            case CacheStatusState::unknown:
+                ::new (&m_known_unavailable_providers)
+                    std::vector<const IBinaryProvider*>(other.m_known_unavailable_providers);
```
Comment (strega-nil-ms) [inline src/vcpkg/binarycaching.cpp:1200]: Why does this use `::new` (as opposed to `new`)?

File: include/vcpkg/binarycaching.h [hunk 4]
```diff
@@ -9,53 +9,131 @@
...
+        void mark_available(const IBinaryProvider* sender) noexcept;
+        void mark_restored() noexcept;
+
+    private:
+        CacheStatusState m_status;
+
+        union
+        {
+            // The set of providers who know they do not have the associated cache entry.
+            // Flat vector because N is tiny.
```
Comment (strega-nil-ms) [inline include/vcpkg/binarycaching.h:66]:
```suggestion
            // Flat vector set because N is tiny.
```

IMO this makes this line less confusing (this line confused me)

File: src/vcpkg/binarycaching.cpp [hunk 5]
```diff
@@ -903,179 +906,490 @@ namespace
...
+        }
+    }
+
+    void CacheStatus::mark_restored() noexcept
+    {
+        switch (m_status)
+        {
+            case CacheStatusState::unknown:
+                m_known_unavailable_providers.~vector();
+                // fallthrough
```
Comment (strega-nil-ms) [inline src/vcpkg/binarycaching.cpp:1376]:
```suggestion
                [[fallthrough]];
```

**Review by strega-nil-ms — APPROVED** [review]
I like this a lot; mostly questions and a few suggestions :)

File: src/vcpkg/dependencies.cpp (hunk 2 above)
Comment (BillyONeal) [reply src/vcpkg/dependencies.cpp:482]: I was getting rid of all the repeated "engaged" checks on optionals in this area

File: src/vcpkg/binarycaching.cpp (hunk 3 above)
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:1200]: Standard library habits .... although something about always using that form for 'True Placement New' has always made sense to me....

Although I haven't bothered with the `static_cast<void*>` part outside of standard library code.

File: src/vcpkg/binarycaching.cpp [hunk 6]
```diff
@@ -409,37 +402,43 @@ namespace
...
 
-                    specs.push_back(action->spec);
-                    auto pkgdir = paths.package_dir(action->spec);
-                    clean_prepare_dir(fs, pkgdir);
-                    pkgdir /= vcpkg::u8path(Strings::concat(*abi.get(), ".zip"));
-                    url_paths.emplace_back(Strings::replace_all(std::string(url_template), "<SHA>", *abi.get()),
-                                           pkgdir);
+                    clean_prepare_dir(fs, paths.package_dir(action.spec));
+                    url_paths.emplace_back(Strings::replace_all(url_template, "<SHA>", *abi.get()),
+                                           make_temp_archive_path(paths.buildtrees, action.spec));
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:440]: Highlighting that this looks like a behavior change: the zip is downloaded into `buildtrees\` instead of `packages\`.

File: include/vcpkg/binarycaching.h [hunk 7]
```diff
@@ -9,53 +9,131 @@
...
-                              std::vector<const Dependencies::InstallPlanAction*>& actions) = 0;
-
-        /// <summary>Requests the result of <c>try_restore()</c> without actually downloading the package. Used by CI to
-        /// determine missing packages.</summary>
-        /// <param name="results_map">InOut map to track the restored packages. Should be initialized to
-        /// <c>{&amp;action, RestoreResult::missing}</c> for all install actions</param>
-        virtual void precheck(
-            const VcpkgPaths& paths,
-            std::unordered_map<const Dependencies::InstallPlanAction*, RestoreResult>& results_map) = 0;
+                              const std::vector<Dependencies::InstallPlanAction>& actions,
```
Comment (ras0219-msft) [inline include/vcpkg/binarycaching.h:92]: Because this api no longer mutates, this could be downgraded to a
```suggestion
                              View<Dependencies::InstallPlanAction> actions,
```

File: include/vcpkg/binarycaching.h [hunk 8]
```diff
@@ -9,53 +9,131 @@
...
+    struct CacheStatus
+    {
+        CacheStatus() noexcept;
+        CacheStatus(const CacheStatus&);
+        CacheStatus(CacheStatus&&) noexcept;
+        CacheStatus& operator=(const CacheStatus&);
+        CacheStatus& operator=(CacheStatus&&) noexcept;
+        ~CacheStatus();
+
+        bool should_attempt_precheck(const IBinaryProvider* sender) const noexcept;
```
Comment (ras0219-msft) [inline include/vcpkg/binarycaching.h:49]: As a general note, these apis feel a bit unfortunate to me (passing `this` to most functions). However, I'm not sure of a cleaner alternative considering the desire to reuse the cache between plans.

The only thing that comes to mind is the full-fortran style of:
```cpp
void IBinaryProvider::prefetch(
    View<InstallAction>,
    View<int> cache_offsets,
    View<enum SharedCacheState>,
    View<enum PrivateCacheState>)
```
Where the owning `BinaryCache` then has:
```cpp
std::unordered_map<PackageSpec, int> offset_map;
std::vector<enum SharedCacheState> shared_state;
std::vector<std::vector<enum PrivateCacheState>> private_states;
std::vector<IBinaryProvider*> providers;
```
with further
```cpp
enum class SharedCacheState : char {
    all_unavailable, // start here, this indicates that every provider so far has confirmed they do not have the package
    any_unknown, // transition to this from all_unavailable if a provider is unable to precheck
    restored, // transition to this if a provider claimed the package
};
enum class PrivateCacheState : char {
    unknown, // I have not prechecked for this package
    unavailable, // I have prechecked for this package and I don't have it
    available, // I have prechecked for this package and I do have it
};
```

Though I don't think this suggestion is sufficiently better to warrant the rewrite.

**Review by ras0219-msft — APPROVED** [review]

File: src/vcpkg/binarycaching.cpp [hunk 9]
```diff
@@ -721,56 +731,45 @@ namespace
...
-                    run_nuget_commandline(cmdline);
-                }();
-
-                Util::erase_remove_if(nuget_refs, [&](const std::pair<PackageSpec, NugetReference>& nuget_ref) -> bool {
+                generate_packages_config(fs, packages_config, nuget_refs);
+                run_nuget_commandline(cmdline);
+                Util::erase_remove_if(nuget_refs, [&](const NuGetPrefetchAttempt& nuget_ref) -> bool {
                     auto nupkg_path =
-                        paths.package_dir(nuget_ref.first) / vcpkg::u8path(nuget_ref.second.id + ".nupkg");
+                        paths.package_dir(nuget_ref.spec) / vcpkg::u8path(nuget_ref.reference.id + ".nupkg");
```
Comment (BillyONeal) [inline src/vcpkg/binarycaching.cpp:746]: This should go into buildtrees

File: src/vcpkg/binarycaching.cpp [hunk 10]
```diff
@@ -602,7 +600,37 @@ namespace
...
+        struct NuGetPrefetchAttempt
+        {
+            PackageSpec spec;
+            NugetReference reference;
+            size_t result_index;
+        };
+
+        static void generate_packages_config(Filesystem& fs,
+                                             const path& packages_config,
+                                             const std::vector<NuGetPrefetchAttempt>& nuget_refs)
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:612]:
```suggestion
                                             const std::vector<NuGetPrefetchAttempt>& prefetches)
```

File: src/vcpkg/binarycaching.cpp [hunk 11]
```diff
@@ -602,7 +600,37 @@ namespace
...
+        {
+            XmlSerializer xml;
+            xml.emit_declaration().line_break();
+            xml.open_tag("packages").line_break();
+
+            for (auto&& nuget_ref : nuget_refs)
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:618]:
```suggestion
            for (auto&& prefetch : prefetches)
```

File: src/vcpkg/binarycaching.cpp (hunk 9 above)
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp:746]: I believe that this is due to the operation of `nuget.exe`, not a choice on the part of vcpkg.

File: src/vcpkg/binarycaching.cpp [hunk 12]
```diff
@@ -359,10 +346,8 @@ namespace
 
                 if (ec)
                 {
-                    vcpkg::printf(Color::warning,
-                                  "Failed to store binary cache %s: %s\n",
-                                  vcpkg::u8string(archive_path),
-                                  ec.message());
+                    vcpkg::printf(
+                        Color::warning, "Failed to store binary cache %s: %s\n", u8string(archive_path), ec.message());
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:350]: There is no other u8string function. We only need to qualify u8path.

File: include/vcpkg/binarycaching.h (hunk 8 above)
Comment (BillyONeal) [reply include/vcpkg/binarycaching.h:49]: I'm not even really convinced that's better -- I think the 'this' use and dual purpose of both uniquely identifying the matching provider(s) and making finding the provider(s) themselves cheap and easy are good properties.

File: src/vcpkg/binarycaching.cpp (hunk 9 above)
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:746]:
```
                    .string_arg("-OutputDirectory")
                    .path_arg(paths.packages)
```

isn't where it comes from?

Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:746]: Oh I see, the client leaves the .nupkg turd around after extraction. Adding comment.

File: src/vcpkg/binarycaching.cpp (hunk 6 above)
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:440]: I think that's a good thing, after discussion made sure they all go into buildtrees

File: src/vcpkg-test/binarycaching.cpp [hunk 13]
```diff
@@ -25,13 +25,13 @@ struct KnowNothingBinaryProvider : IBinaryProvider
 
     virtual void push_success(const VcpkgPaths&, const Dependencies::InstallPlanAction&) const override { }
     virtual void prefetch(const VcpkgPaths&,
-                          const std::vector<Dependencies::InstallPlanAction>&,
-                          const std::vector<CacheStatus*>&) const override
+                          View<const Dependencies::InstallPlanAction>,
+                          View<CacheStatus* const>) const override
```
Comment (ras0219-msft) [inline src/vcpkg-test/binarycaching.cpp:29]:
```suggestion
                          View<Dependencies::InstallPlanAction>,
                          View<CacheStatus*>) const override
```
`View<T>` implies const (it is just `Span<const T>`)

File: src/vcpkg-test/binarycaching.cpp [hunk 14]
```diff
@@ -25,13 +25,13 @@ struct KnowNothingBinaryProvider : IBinaryProvider
...
     {
     }
     virtual void precheck(const VcpkgPaths&,
-                          const std::vector<Dependencies::InstallPlanAction>&,
-                          const std::vector<CacheStatus*>& cache_status) const override
+                          View<const Dependencies::InstallPlanAction>,
+                          View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg-test/binarycaching.cpp:34]:
```suggestion
                          View<Dependencies::InstallPlanAction>,
                          View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 15]
```diff
@@ -247,8 +247,8 @@ namespace
         }
 
         void prefetch(const VcpkgPaths& paths,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:251]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 16]
```diff
@@ -362,8 +362,8 @@ namespace
         }
 
         void precheck(const VcpkgPaths& paths,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:366]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 17]
```diff
@@ -415,8 +415,8 @@ namespace
         void push_success(const VcpkgPaths&, const Dependencies::InstallPlanAction&) const override { }
 
         void prefetch(const VcpkgPaths& paths,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:419]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 18]
```diff
@@ -473,8 +473,8 @@ namespace
         }
 
         void precheck(const VcpkgPaths&,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:477]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 19]
```diff
@@ -629,8 +629,8 @@ namespace
         }
 
         void prefetch(const VcpkgPaths& paths,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:633]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 20]
```diff
@@ -869,8 +871,8 @@ namespace
         }
 
         void precheck(const VcpkgPaths&,
-                      const std::vector<Dependencies::InstallPlanAction>&,
-                      const std::vector<CacheStatus*>&) const override
+                      View<const Dependencies::InstallPlanAction>,
+                      View<CacheStatus* const>) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:875]:
```suggestion
                      View<Dependencies::InstallPlanAction>,
                      View<CacheStatus*>) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 21]
```diff
@@ -935,8 +937,8 @@ namespace
         }
 
         void prefetch(const VcpkgPaths& paths,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:941]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 22]
```diff
@@ -1011,8 +1013,8 @@ namespace
         }
 
         void precheck(const VcpkgPaths&,
-                      const std::vector<Dependencies::InstallPlanAction>& actions,
-                      const std::vector<CacheStatus*>& cache_status) const override
+                      View<const Dependencies::InstallPlanAction> actions,
+                      View<CacheStatus* const> cache_status) const override
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:1017]:
```suggestion
                      View<Dependencies::InstallPlanAction> actions,
                      View<CacheStatus*> cache_status) const override
```

File: src/vcpkg/binarycaching.cpp [hunk 23]
```diff
@@ -1144,9 +1146,8 @@ namespace vcpkg
         }
     }
 
-    static std::vector<CacheStatus*> build_cache_status_vector(
-        const std::vector<Dependencies::InstallPlanAction>& actions,
-        std::unordered_map<std::string, CacheStatus>& status)
+    static std::vector<CacheStatus*> build_cache_status_vector(View<const Dependencies::InstallPlanAction> actions,
+                                                               std::unordered_map<std::string, CacheStatus>& status)
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:1150]:
```suggestion
    static std::vector<CacheStatus*> build_cache_status_vector(View<Dependencies::InstallPlanAction> actions,
                                                               std::unordered_map<std::string, CacheStatus>& status)
```

File: src/vcpkg/binarycaching.cpp [hunk 24]
```diff
@@ -1161,7 +1162,7 @@ namespace vcpkg
         return results;
     }
 
-    void BinaryCache::prefetch(const VcpkgPaths& paths, const std::vector<Dependencies::InstallPlanAction>& actions)
+    void BinaryCache::prefetch(const VcpkgPaths& paths, View<const Dependencies::InstallPlanAction> actions)
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:1165]:
```suggestion
    void BinaryCache::prefetch(const VcpkgPaths& paths, View<Dependencies::InstallPlanAction> actions)
```

File: src/vcpkg/binarycaching.cpp [hunk 25]
```diff
@@ -1178,7 +1179,7 @@ namespace vcpkg
     }
 
     std::vector<CacheAvailability> BinaryCache::precheck(const VcpkgPaths& paths,
-                                                         const std::vector<Dependencies::InstallPlanAction>& actions)
+                                                         View<const Dependencies::InstallPlanAction> actions)
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:1182]:
```suggestion
                                                         View<Dependencies::InstallPlanAction> actions)
```

**Review by ras0219-msft — APPROVED** [review]

**Comment by dg0yt** [comment]
PS: Obviously there is a cache check before uploading. The mentioned osc build uploaded only 178 packages from 522 builds. 
Example build without upload:
~~~
Building package kenlm[core]:x64-osx... done
Installing package kenlm[core]:x64-osx...
Installing package kenlm[core]:x64-osx... done
Elapsed time for package kenlm:x64-osx: 47.81 s
~~~
So the impact of check-before-build is even even less, given the absolute number of checks (early, before build, after build).

**Comment by BillyONeal** [comment]
> IIUC this change has an unexpected negative side effect when multiple PR are staged a short time after merges to `master`.
> 
> * When the `vcpkg ci` command starts, it determines which ports to install by checking for cache misses.
>   For some time after a merge to master, all PRs can have _many cache misses due to the changes in master_, not in the PR.
> * With this PR, the _early cache misses are marked as unavailable_.
> * Now the `vcpkg ci` commands starts to install (build) the missing ports. This takes a while.
> * Marked as unavailable, the _early cache misses will not be checked again at install time_ to see if they became available in the meantime.
> 
> So multiple, unrelated PRs will build the same ports for the same master configuration in parallel.

That was true before this change as well:

https://github.com/microsoft/vcpkg-tool/blob/e5242a3db56390b640bef4e85b2d18ab7719204e/src/vcpkg/binarycaching.cpp#L477-L485

I agree that some means to "share" this load would be an improvement; I didn't go there on this PR because it was big enough only resolving the "hammer Blob Storage so much that we blow fuses" problem :)

> Extreme example: My Azure Piplelines markup PR [microsoft/vcpkg#19104](https://github.com/microsoft/vcpkg/pull/19104) which doesn't touch a single port or vcpkg script.
> On osx, it built 522 ports, taking more than 4 hours:
> https://dev.azure.com/vcpkg/public/_build/results?buildId=56644&view=logs&j=2644860b-263e-5de8-c39e-3f62c017aad0&t=0c846259-ebdc-51ce-09bf-fc9ed3a241c7
> Osx started early after push. x64-windows-static suffered, too. The other targets waited much longer for a host to run the build, so they were happy to find more cache hits when they started.
> 
> I would propose either to not mark binary artifacts as unavailable (which could double the number of cache request in the worst case), or to add a time-to-live to the cache results.

Time To Live is a good idea. I've been wondering how we could fix it to avoid repeated work like you describe but had not come up with good ideas before now.

> PS: Obviously there is a cache check before uploading. The mentioned osc build uploaded only 178 packages from 522 builds.
> Example build without upload:
> 
> ```
> Building package kenlm[core]:x64-osx... done
> Installing package kenlm[core]:x64-osx...
> Installing package kenlm[core]:x64-osx... done
> Elapsed time for package kenlm:x64-osx: 47.81 s
> ```
> 
> So the impact of check-before-build is even even less, given the absolute number of checks (early, before build, after build).

There actually isn't a cache check before uploading :)

**Comment by dg0yt** [comment]
Thanks for taking the time to help me understand.

> That was true before this change as well:

You are pointing at `try_restore()`. But I'm concerned about the changes to the implementations of `precheck()`:
- Before the PR, they returned a map of (action, state). *No side effect.* The state was used to collect ports to install. When arriving at install, the state would be checked again. 
- Now they have the *side effect* of changing status on the `BinaryCache` object by calling `cache_status[idx]->mark_unavailable(this);`. (These lines of code are new, and they lead me here via Blame on Github.)
  IIUC this status is now reused when arriving at install.

**Comment by dg0yt** [comment]
Looks like your current pistache removal PR gives another extreme example: No dependencies, but hundreds of packages are being rebuilt, taking hours, and other PRs pile up waiting for CI machines. It is really too much at the moment.

**Comment by BillyONeal** [comment]
> Thanks for taking the time to help me understand.
> 
> > That was true before this change as well:
> 
> You are pointing at `try_restore()`. But I'm concerned about the changes to the implementations of `precheck()`:
> 
> * Before the PR, they returned a map of (action, state). _No side effect._ The state was used to collect ports to install. When arriving at install, the state would be checked again.
> * Now they have the _side effect_ of changing status on the `BinaryCache` object by calling `cache_status[idx]->mark_unavailable(this);`. (These lines of code are new, and they lead me here via Blame on Github.)
>   IIUC this status is now reused when arriving at install.

Before this same side effect was in prefetch rather than precheck. I highlighted the try_restore line to indicate that no download attempts occur while builds are running; there’s no download code there :).

Before the side effect was in m_restored on prefetch, now the side effect is moved earlier to precheck; the difference is that we no longer try to prefetch all 1600 things 2 seconds after prefetch told us they didn’t exist.
